

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>align package &mdash; align 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> align
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">align package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-align.algs">align.algs module</a><ul>
<li><a class="reference internal" href="#algs">algs</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-align">Module contents</a></li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">align</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>align package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/align.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="align-package">
<h1>align package<a class="headerlink" href="#align-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-align.algs">
<span id="align-algs-module"></span><h2>align.algs module<a class="headerlink" href="#module-align.algs" title="Permalink to this headline">¶</a></h2>
<div class="section" id="algs">
<h3>algs<a class="headerlink" href="#algs" title="Permalink to this headline">¶</a></h3>
<p>The core algorithm module of my BMI 203 Project 1</p>
<dl class="py class">
<dt id="align.algs.NeedlemanWunsch">
<em class="property">class </em><code class="sig-prename descclassname">align.algs.</code><code class="sig-name descname">NeedlemanWunsch</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">substitution_matrix_file</span></em>, <em class="sig-param"><span class="n">gap_open</span></em>, <em class="sig-param"><span class="n">gap_extend</span></em>, <em class="sig-param"><span class="n">overlap_alignment</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#align.algs.NeedlemanWunsch" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#align.algs.PairwiseAligner" title="align.algs.PairwiseAligner"><code class="xref py py-class docutils literal notranslate"><span class="pre">align.algs.PairwiseAligner</span></code></a></p>
<dl class="simple">
<dt>This class is an implenetation of the NeedlemanWunsch pairwise</dt><dd><p>global aligner algorithm and inherets from the PairwiseAligner
class. This class can be used to align any two protein sequences.</p>
</dd>
</dl>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>substitution_matrix_file</strong><span class="classifier">str</span></dt><dd><p>path/filename of substitution matrix for NeedlemanWunsch</p>
</dd>
<dt><strong>gap_open</strong><span class="classifier">float</span></dt><dd><p>gap opening penalty for NeedlemanWunsch</p>
</dd>
<dt><strong>gap_extend</strong><span class="classifier">float</span></dt><dd><blockquote>
<div><p>gap extension penalty for NeedlemanWunsch</p>
</div></blockquote>
<dl class="simple">
<dt>overlap_alignment<span class="classifier">bool, default=False</span></dt><dd><p>whether or not to allow variable start stop for overlap
alignment, which does not assume the reads start at the
same resiude.</p>
</dd>
</dl>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>seq1</strong><span class="classifier">str</span></dt><dd><blockquote>
<div><p>sequence from fasta1 for NeedlemanWunsch</p>
</div></blockquote>
<dl class="simple">
<dt>seq2<span class="classifier">str</span></dt><dd><p>sequence from fasta2 for NeedlemanWunsch</p>
</dd>
<dt>header1<span class="classifier">str</span></dt><dd><p>fasta header from fasta1 for NeedlemanWunsch</p>
</dd>
<dt>header2<span class="classifier">str</span></dt><dd><p>fasta header from fasta2 for NeedlemanWunsch</p>
</dd>
<dt><a href="#id1"><span class="problematic" id="id2">alignment_</span></a><span class="classifier">tuple</span></dt><dd><p>Tuple of format (seq1_alignment, seq2_alignment) for NeedlemanWunsch</p>
</dd>
<dt><a href="#id3"><span class="problematic" id="id4">alignment_score_</span></a><span class="classifier">float</span></dt><dd><p>Score of alignment from algorithm for NeedlemanWunsch</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(fasta1, fasta2[, return_alignment])</p></td>
<td><p>This is a method that must be overwritten when inhereting from this class.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt id="align.algs.PairwiseAligner">
<em class="property">class </em><code class="sig-prename descclassname">align.algs.</code><code class="sig-name descname">PairwiseAligner</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">substitution_matrix_file</span></em>, <em class="sig-param"><span class="n">gap_open</span></em>, <em class="sig-param"><span class="n">gap_extend</span></em><span class="sig-paren">)</span><a class="headerlink" href="#align.algs.PairwiseAligner" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Paiwise Alignment Base Class</p>
<blockquote>
<div><p>This is the base (parent) class for all pairwise alignment algorithms
implemented in this API. All pariwise alignment algorithm classes
inheret from this base class. This class can be used to create
any pairwise alignment algorithm class.</p>
</div></blockquote>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>substitution_matrix_file</strong><span class="classifier">str</span></dt><dd><p>path/filename of substitution matrix</p>
</dd>
<dt><strong>gap_open</strong><span class="classifier">float</span></dt><dd><p>gap opening penalty (should be positive value, think abs value)</p>
</dd>
<dt><strong>gap_extend</strong><span class="classifier">float</span></dt><dd><p>gap extension penalty (should be postivie value, think abs value)</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes</dt>
<dd class="field-even"><dl>
<dt><strong>seq1</strong><span class="classifier">str</span></dt><dd><blockquote>
<div><p>sequence from fasta1</p>
</div></blockquote>
<dl class="simple">
<dt>seq2<span class="classifier">str</span></dt><dd><p>sequence from fasta2</p>
</dd>
<dt>header1<span class="classifier">str</span></dt><dd><p>fasta header from fasta1</p>
</dd>
<dt>header2<span class="classifier">str</span></dt><dd><p>fasta header from fasta2</p>
</dd>
<dt><a href="#id5"><span class="problematic" id="id6">alignment_</span></a><span class="classifier">tuple</span></dt><dd><p>Tuple of format (seq1_alignment, seq2_alignment)</p>
</dd>
<dt><a href="#id7"><span class="problematic" id="id8">alignment_score_</span></a><span class="classifier">float</span></dt><dd><p>Score of alignment from algorithm</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#align.algs.PairwiseAligner.align" title="align.algs.PairwiseAligner.align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code></a>(fasta1, fasta2[, return_alignment])</p></td>
<td><p>This is a method that must be overwritten when inhereting from this class.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="align.algs.PairwiseAligner.align">
<code class="sig-name descname">align</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fasta1</span></em>, <em class="sig-param"><span class="n">fasta2</span></em>, <em class="sig-param"><span class="n">return_alignment</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#align.algs.PairwiseAligner.align" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a method that must be overwritten when inhereting from
this class. for NeedlemanWunsch</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fasta1</strong> (<em>str</em>) – Filename string of first fasta for alignment</p></li>
<li><p><strong>fasta2</strong> (<em>str</em>) – Filename string of second fasta for alignment</p></li>
<li><p><strong>return_alignment</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether or not to return the alignmenti if True then
the alignment is returned and if False then None
is returned</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>alignment</strong> – Tuple of format (seq1_alignment, seq2_alignment)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="align.algs.SmithWaterman">
<em class="property">class </em><code class="sig-prename descclassname">align.algs.</code><code class="sig-name descname">SmithWaterman</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">substitution_matrix_file</span></em>, <em class="sig-param"><span class="n">gap_open</span></em>, <em class="sig-param"><span class="n">gap_extend</span></em><span class="sig-paren">)</span><a class="headerlink" href="#align.algs.SmithWaterman" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#align.algs.PairwiseAligner" title="align.algs.PairwiseAligner"><code class="xref py py-class docutils literal notranslate"><span class="pre">align.algs.PairwiseAligner</span></code></a></p>
<p>This class is an implenetation of the Smith Waterman pairwise
local aligner algorithm and inherets from the PairwiseAligner
class. This class can be used to align any two protein sequences
for SmithWaterman.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>substitution_matrix_file</strong> (<em>str</em>) – path/filename of substitution matrix</p></li>
<li><p><strong>gap_open</strong> (<em>float</em>) – gap opening penalty</p></li>
<li><p><strong>gap_extend</strong> (<em>float</em>) – gap extension penalty</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt id="align.algs.SmithWaterman.seq1">
<code class="sig-name descname">seq1</code><a class="headerlink" href="#align.algs.SmithWaterman.seq1" title="Permalink to this definition">¶</a></dt>
<dd><p>sequence from fasta1</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="align.algs.SmithWaterman.seq2">
<code class="sig-name descname">seq2</code><a class="headerlink" href="#align.algs.SmithWaterman.seq2" title="Permalink to this definition">¶</a></dt>
<dd><p>sequence from fasta2</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="align.algs.SmithWaterman.header1">
<code class="sig-name descname">header1</code><a class="headerlink" href="#align.algs.SmithWaterman.header1" title="Permalink to this definition">¶</a></dt>
<dd><p>fasta header from fasta1</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="align.algs.SmithWaterman.header2">
<code class="sig-name descname">header2</code><a class="headerlink" href="#align.algs.SmithWaterman.header2" title="Permalink to this definition">¶</a></dt>
<dd><p>fasta header from fasta2</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="align.algs.SmithWaterman.alignment_">
<code class="sig-name descname">alignment_</code><a class="headerlink" href="#align.algs.SmithWaterman.alignment_" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple of format (seq1_alignment, seq2_alignment)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="align.algs.SmithWaterman.alignment_score_">
<code class="sig-name descname">alignment_score_</code><a class="headerlink" href="#align.algs.SmithWaterman.alignment_score_" title="Permalink to this definition">¶</a></dt>
<dd><p>Score of alignment from algorithm</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(fasta1, fasta2[, return_alignment])</p></td>
<td><p>This is a method that must be overwritten when inhereting from this class.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>
<div class="section" id="module-align">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-align" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Parker Grosjean.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>